jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt

```diff
@@ -0,0 +1,305 @@
+import com.jetbrains.rider.debugger.actions.utils.OptionsUtil
+import com.jetbrains.rider.model.debuggerWorker.DotNetCoreExeStartInfo
+import com.jetbrains.rider.model.debuggerWorker.DotNetCoreInfo
+import com.jetbrains.rider.model.debuggerWorker.DotNetDebuggerSessionModel
+import com.jetbrains.rider.model.debuggerWorkerConnectionHelperModel
+import com.jetbrains.rider.projectView.solution
+import com.jetbrains.rider.run.ConsoleKind
+import com.jetbrains.rider.run.IDebuggerOutputListener
+import com.jetbrains.rider.run.bindToSettings
+import org.jetbrains.concurrency.AsyncPromise
+import org.jetbrains.concurrency.Promise
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.trace
+import software.aws.toolkits.jetbrains.services.lambda.execution.local.SamDebugSupport
+import software.aws.toolkits.jetbrains.services.lambda.execution.local.SamRunningState
+import software.aws.toolkits.resources.message
+import java.io.File
+import java.io.OutputStream
+import java.util.Timer
+import kotlin.concurrent.schedule
+
+/**
+ * Debugger is launched via SAM CLI command: '$ sam local invoke' and pass debugger parameters --debugger-path and --debugger-args.
+ * Under SAM CLI this command is translated into:
+ * - volume that is mounted in Docker to launch debugger (--debugger-path)
+ * - additional debugger arguments that are applied when launching debugger inside Docker (--debugger-args):
+ *    '$ dotnet <debug_args_list> MockBootstraps.dll --debugger-spin-wait'
+ *
+ * Rider uses it's own Debugger Worker process that run under .NET with extra parameters to configure:
+ * - path to runtime
+ * - path to DLL
+ * - debugger mode (client/server)
+ * - frontend port
+ * - backend port
+ *
+ * To launch Rider Debugger under Docker we have created a simple dotnet core app that launch Debugger Worker instance
+ * using Mono runtime that is taken from Rider SDK (runtime.sh).
+ * This app is hosted in github repo - https://github.com/JetBrains/JetBrains.Rider.Debugger.Launcher and is
+ * delivered via GutHub releases with packed output.
+ * This output is copied to the same Rider SDK path that is mounted in Docker (app should be placed under "Amazon.Runner" folder).
+ *
+ * After Debugger Worker process is started on Docker instance it initializes ports and process to run - MockBootstraps.dll
+ * (the internal compiled with lambda code)
+ */
+class DotNetSamDebugSupport : SamDebugSupport {
+    companion object {
+        private val logger = getLogger<DotNetSamDebugSupport>()
+
+        private const val debuggerDir = "/tmp/lambci_debug_files"
+        private const val dotnetCorCliPath = "/var/lang/bin/dotnet"
+        private const val mockBootstrapsPath = "/var/runtime/MockBootstraps.dll"
+    }
+
+    /**
+     * Check whether the Rider Debugger Worker Launcher app (that is required to run Debugger) is downloaded into Rider SDK.
```

Comment:
I am trying to wrap my head around this and the gradle changes.

Is `Amazon.Runner` shipped inside of the plugin? I am just a little confused based on the "Rider SDK" comment.

Reply:
Sorry about confusion. I've tried to fit some details in the description that might be not straightforward. The initial idea was updated to ship JetBrains.Rider.Debugger.Launcher app inside Rider SDK. Updated the comment accordingly. Please feel free to comment on the new one if there are any questions left for a new approach.